### PR TITLE
[Indexer] Stream full blocks instead of transactions

### PR DESCRIPTION
**Pull request by bowenyang007**
### Description
Stream full blocks in the format of
```
FIRE INIT aptos-node 0.1.0 aptos 0 0
FIRE BLOCK_START <height>
FIRE TRX <version> <transaction proto base64>
FIRE TRX <version> <transaction proto base64>
…
FIRE BLOCK_END
```
Added validation for the transactions to ensure correctness before we stream the data out. If anything is wrong with any of the transactions within a block, we fail the entire block and do not stream anything. 

### Test Plan
cargo test
ensure `cargo run -p aptos -- node run-local-testnet --with-faucet` works as expected

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/2821)

### CODE REVIEW

File: sf-stream/src/runtime.rs [hunk 1]
```diff
@@ -109,80 +93,83 @@ impl SfStreamer {
...
-            block_height: block_info.block_height,
-            current_epoch: epoch,
+            current_block_height: block_event.height(),
+            current_epoch: block_event.epoch(),
             mp_sender: mp_client_sender,
         }
     }
 
     pub async fn start(&mut self) {
+        println!("DMLOG INIT aptos v1");
```
Comment (bowenyang007) [inline sf-stream/src/runtime.rs:104]: @CapCap halp!

File: sf-stream/src/runtime.rs [hunk 2]
```diff
@@ -73,33 +75,15 @@ impl SfStreamer {
         mp_client_sender: Option<MempoolClientSender>,
     ) -> Self {
         let resolver = Arc::new(context.move_resolver().unwrap());
-        let latest = context.get_latest_ledger_info().unwrap();
-        let block_info = context
-            .get_block_info(starting_version, latest.ledger_version.0)
+        let (_block_start_version, _block_last_versionn, block_event) = context
```
Comment (CapCap) [inline sf-stream/src/runtime.rs:78]: does this work for genesis too now?

File: sf-stream/src/runtime.rs [hunk 3]
```diff
@@ -109,80 +93,86 @@ impl SfStreamer {
...
-                    "[sf-stream] got {} transactions from {} to {} [{}]",
+                    "[sf-stream] got {} transactions from {} to {} [version on last actual transaction {}]",
                     transactions.len(),
-                    self.current_version,
-                    self.current_version + transactions.len() as u64,
+                    block_start_version,
+                    block_last_version,
                     transactions.last().map(|txn| txn.version).unwrap_or(0)
                 );
+                let mut block_timestamp = None;
```
Comment (CapCap) [inline sf-stream/src/runtime.rs:157]: why not  this?
```
let block_timestamp =  self.context
                                .get_block_timestamp(block_start_version)
                                .unwrap_or_else(|_| {
                                    panic!(
                                        "Could not get timestamp for version {}",
                                        block_start_version
                                    )
                                })
```

File: sf-stream/src/runtime.rs [hunk 4]
```diff
@@ -191,43 +181,101 @@ impl SfStreamer {
...
 
-    pub fn convert_transaction(&mut self, transaction: Transaction) -> extractor::Transaction {
-        self.maybe_update_from_block_metadata(&transaction);
-        convert_transaction(&transaction, self.block_height, self.current_epoch)
+    /// We can consider a block height as valid if these conditions are met:
+    /// 1. first (and only first) transaction is a block metadata or genesis 2. versions are monotonically increasing 3. start and end versions match block boundaries
+    /// Return error if the block is not valid. Panic if there's anything wrong with encoding a transaction.
+    fn print_block_with_validation(
+        &self,
+        converted_txns: &Vec<TransactionPB>,
```
Comment (CapCap) [inline sf-stream/src/runtime.rs:215]: no reason to build up a vec of these; all validations can be done much more simply as we convert them

File: sf-stream/src/runtime.rs [hunk 5]
```diff
@@ -191,43 +181,101 @@ impl SfStreamer {
...
         let mut buf = vec![];
         transaction.encode(&mut buf).unwrap_or_else(|_| {
             panic!(
                 "Could not convert protobuf transaction to bytes '{:?}'",
                 transaction
             )
         });
-        let pb_b64 = &base64::encode(buf);
-        println!("DMLOG TRX {}", pb_b64);
+        println!("FIRE TRX {}", base64::encode(buf));
```
Comment (CapCap) [inline sf-stream/src/runtime.rs:278]: why we move from `DMLOG` to `FIRE`?

File: sf-stream/src/runtime.rs (hunk 2 above)
Comment (bowenyang007) [reply sf-stream/src/runtime.rs:78]: yes

File: sf-stream/src/runtime.rs (hunk 4 above)
Comment (bowenyang007) [reply sf-stream/src/runtime.rs:215]: hmm fair.

File: sf-stream/src/runtime.rs [hunk 6]
```diff
@@ -109,125 +92,162 @@ impl SfStreamer {
...
-        convert_transaction(&transaction, self.block_height, self.current_epoch)
+    /// First, and only first, transaction in a block has to be bmt or genesis
+    fn validate_transaction_type(&self, is_first_txn: bool, transaction: &Transaction) -> bool {
+        let is_bm_or_genesis = match transaction {
+            Transaction::BlockMetadataTransaction(_) => true,
+            Transaction::GenesisTransaction(_) => true,
+            Transaction::UserTransaction(_) => false,
+            Transaction::StateCheckpointTransaction(_) => false,
+            Transaction::PendingTransaction(_) => false,
+        };
```
Comment (CapCap) [inline sf-stream/src/runtime.rs:238]:
```
        let is_bm_or_genesis = match transaction {
            Transaction::BlockMetadataTransaction(_) => true,
            Transaction::GenesisTransaction(_) => true,
            _ => false,
        };
```

Comment (bowenyang007) [reply sf-stream/src/runtime.rs:238]: ah love this.

Comment (bowenyang007) [reply sf-stream/src/runtime.rs:238]: xclippy gave me something even better lol 
```
matches!(
            transaction,
            Transaction::BlockMetadataTransaction(_) | Transaction::GenesisTransaction(_)
        );
```

Comment (CapCap) [reply sf-stream/src/runtime.rs]: nice!

**Review by CapCap — APPROVED** [review]

**Review by CapCap — APPROVED** [review]

**Pull request merged**